IllegalStateException: WRITER when using PrintWriter from ctx.res()

From Javalin 5.x onwards the following error message is written to the logs when using PrintWriter from Context.res().getWriter()

> io.javalin.Javalin - Exception occurred while servicing http-request
> java.lang.IllegalStateException: WRITER

No such log messages where observed for Javalin < 5

**Reproduce with**
```
final var app = Javalin.create();
app.get("/oops", ctx -> {
    ctx.res().setStatus(200);
    ctx.res().setContentType("text/plain");
    final var writer = ctx.res().getPrintWriter();
    writer.print("test");
    writer.close();
});
app.start();
```

**Additional context**
Based on the stack trace below, Javalin uses ServletResponse#getOutputStream() for OutputStream compression. This is not allowed when ServletResponse#getWriter() has been called on this response object.

> io.javalin.Javalin - Exception occurred while servicing http-request
> java.lang.IllegalStateException: WRITER
> at org.eclipse.jetty.server.Response.getOutputStream(Response.java:868)
> at io.javalin.compression.CompressedOutputStream.<init>(CompressedStream.kt:16)
> at io.javalin.http.servlet.JavalinServletContext$outputStreamWrapper$2.invoke(JavalinServletContext.kt:111)
> at io.javalin.http.servlet.JavalinServletContext$outputStreamWrapper$2.invoke(JavalinServletContext.kt:111)
>

This is the full stack trace from our logs
```

[2022-11-17 13:20:54:753 +0000] [qtp837108062-16] [kvo-568c987ccb-knf6w] [CONFIDENTIAL] [null] [ERROR] io.javalin.Javalin - Exception occurred while servicing http-request
java.lang.IllegalStateException: WRITER
        at org.eclipse.jetty.server.Response.getOutputStream(Response.java:868)
        at io.javalin.compression.CompressedOutputStream.<init>(CompressedStream.kt:16)
        at io.javalin.http.servlet.JavalinServletContext$outputStreamWrapper$2.invoke(JavalinServletContext.kt:111)
        at io.javalin.http.servlet.JavalinServletContext$outputStreamWrapper$2.invoke(JavalinServletContext.kt:111)
        at kotlin.SynchronizedLazyImpl.getValue(LazyJVM.kt:74)
        at io.javalin.http.servlet.JavalinServletContext.getOutputStreamWrapper(JavalinServletContext.kt:111)
        at io.javalin.http.servlet.JavalinServletContext.outputStream(JavalinServletContext.kt:112)
        at io.javalin.http.servlet.JavalinServlet.writeResponseAndLog(JavalinServlet.kt:90)
        at io.javalin.http.servlet.JavalinServlet.handleSync(JavalinServlet.kt:50)
        at io.javalin.http.servlet.JavalinServlet.service(JavalinServlet.kt:34)
        at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:587)
        at io.javalin.jetty.JavalinJettyServlet.service(JavalinJettyServlet.kt:58)
        at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:587)
        at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:764)
        at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:529)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:221)
        at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1571)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:221)
        at io.javalin.jetty.JettyServer$start$wsAndHttpHandler$1.doHandle(JettyServer.kt:57)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:176)
        at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:484)
        at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1544)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:174)
        at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1302)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:129)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:122)
        at org.eclipse.jetty.server.Server.handle(Server.java:563)
        at org.eclipse.jetty.server.HttpChannel.lambda$handle$0(HttpChannel.java:505)
        at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:762)
        at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:497)
        at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:282)
        at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:314)
        at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:100)
        at org.eclipse.jetty.io.SelectableChannelEndPoint$1.run(SelectableChannelEndPoint.java:53)
        at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.runTask(AdaptiveExecutionStrategy.java:421)
        at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.consumeTask(AdaptiveExecutionStrategy.java:390)
        at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.tryProduce(AdaptiveExecutionStrategy.java:277)
        at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.lambda$new$0(AdaptiveExecutionStrategy.java:139)
        at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:411)
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:933)
        at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1077)
        at java.base/java.lang.Thread.run(Thread.java:833)
```

--- a/javalin/src/main/java/io/javalin/http/servlet/JavalinServlet.kt
+++ b/javalin/src/main/java/io/javalin/http/servlet/JavalinServlet.kt
@@ -94,8 +94,8 @@ class JavalinServlet(val cfg: JavalinConfig) : HttpServlet() {
     private fun JavalinServletContext.writeResponseAndLog() {
         try {
             if (responseWritten.getAndSet(true)) return // prevent writing more than once, it's required because timeout listener can terminate the flow at any time
-            outputStream().use { outputStream ->
-                resultInputStream()?.use { resultStream ->
+            resultInputStream()?.use { resultStream ->
+                outputStream().use { outputStream ->
                     val etagWritten = ETagGenerator.tryWriteEtagAndClose(cfg.http.generateEtags, this, resultStream)
                     if (!etagWritten) resultStream.copyTo(outputStream, 4096)
                 }

--- a/javalin/src/main/java/io/javalin/http/servlet/JavalinServlet.kt
+++ b/javalin/src/main/java/io/javalin/http/servlet/JavalinServlet.kt
@@ -95,10 +95,8 @@ class JavalinServlet(val cfg: JavalinConfig) : HttpServlet() {
         try {
             if (responseWritten.getAndSet(true)) return // prevent writing more than once, it's required because timeout listener can terminate the flow at any time
             resultInputStream()?.use { resultStream ->
-                outputStream().use { outputStream ->
-                    val etagWritten = ETagGenerator.tryWriteEtagAndClose(cfg.http.generateEtags, this, resultStream)
-                    if (!etagWritten) resultStream.copyTo(outputStream, 4096)
-                }
+                val etagWritten = ETagGenerator.tryWriteEtagAndClose(cfg.http.generateEtags, this, resultStream)
+                if (!etagWritten) resultStream.copyTo(outputStream(), 4096)
             }
             cfg.pvt.requestLogger?.handle(this, executionTimeMs())
         } catch (throwable: Throwable) {

--- a/javalin/src/main/java/io/javalin/http/servlet/JavalinServlet.kt
+++ b/javalin/src/main/java/io/javalin/http/servlet/JavalinServlet.kt
@@ -104,6 +104,7 @@ class JavalinServlet(val cfg: JavalinConfig) : HttpServlet() {
         } catch (throwable: Throwable) {
             exceptionMapper.handleUnexpectedThrowable(res(), throwable) // handle any unexpected error, e.g. write failure
         } finally {
+            if (outputStreamWrapper.isReady) outputStream().close() // close initialized output wrappers
             if (isAsync()) req().asyncContext.complete() // guarantee completion of async context to eliminate the possibility of hanging connections
         }
     }

--- a/javalin/src/main/java/io/javalin/http/servlet/JavalinServletContext.kt
+++ b/javalin/src/main/java/io/javalin/http/servlet/JavalinServletContext.kt
@@ -108,7 +108,7 @@ class JavalinServletContext(
     private val queryParams by lazy { super.queryParamMap() }
     override fun queryParamMap(): Map<String, List<String>> = queryParams
 
-    private val outputStreamWrapper by lazy { CompressedOutputStream(compressionStrategy, this) }
+    internal val outputStreamWrapper by lazy { CompressedOutputStream(compressionStrategy, this) }
     override fun outputStream(): ServletOutputStream = outputStreamWrapper
 
     override fun redirect(location: String, status: HttpStatus) {

